Add Doxygen documentation infrastructure

glopesdev opened the pull request:
* Doxygen version 1.12.0 and the Doxygen Awesome CSS theme were used.
* Root favicon and logo added for consistency with other Harp documentation websites.
* Triple slash added to existing function documentation comments so it gets automatically picked up by Doxygen.
* `OPTIMIZE_OUTPUT_FOR_C` flag was set to adjust generated docs terminology.

Following merge to main the new layout will be published.

------------

Documentation will automatically publish to GitHub Pages when:

* A release is made
* A manual workflow dispatch is initiated and publishing is enabled
* Any push to any branch is made when `CONTINUOUS_DOCUMENTATION` is enabled (intended for use in forks for development purposes, but this is not enforced.)

Preview at: https://glopesdev.github.io/core.atxmega/

glopesdev committed:
Add Doxygen documentation infrastructure

Moved the .gitignore file to the root of the repository to facilitate
central editing.

glopesdev committed:
Use triple slash in function comment blocks

This is to ensure Doxygen can pick up these descriptions into generated
documentation.

glopesdev committed:
Add documentation CI/CD workflow

PathogenDavid commented:
For the benefit of anyone else who happens to notice that the [globals index](https://glopesdev.github.io/core.atxmega/globals.html) on the ATxmega core isn't super great compared to [the same page in the Pico core](https://glopesdev.github.io/core.pico/globals.html):

Doxygen has a [hard-coded cutoff](https://github.com/doxygen/doxygen/blob/f134637b01aa8479bc853c51d6159d51e13d05f4/src/index.cpp#L3333) where it splits those indices across multiple files and it's absurdly low at just 200 members. (This limit was set back in 2005 when dinosaurs still roamed the earth.)

Also worth noting that `ALPHABETICAL_INDEX` is completely unrelated to these pages. That corresponds to [this index](https://glopesdev.github.io/core.pico/classes.html) instead. (Quite annoyingly, this file [doesn't seem to have a limit](https://clang.llvm.org/doxygen/classes.html) and even intelligently puts `_` at the end.)

There's actually [an issue](https://www.github.com/doxygen/doxygen/issues/9182) and [corresponding PR](https://www.github.com/doxygen/doxygen/pull/9194) to improve this, but both seem to have stalled out. (To my dismay the PR unfortunately also doesn't let you disable the feature entirely and caps the maximum at 3,000.)

PathogenDavid reviewed (commented):
Everything I mentioned in https://github.com/harp-tech/core.pico/pull/56#pullrequestreview-2929042344 also applies to this PR.

PathogenDavid commented on Firmware/Library/hwbp_core.h:
```diff
@@ -18,102 +18,102 @@ typedef struct {
 } timestamp_t;
 
-// Used to define the clock direction default of the device.
+/// Used to define the clock direction default of the device.
```
I don't have a strong opinion one way or another, but in my experience `//!` is normative for C/C++ projects.

Not sure if @filcarv has a preference.

glopesdev replied:
Keeping it as-is for now, but we can always change later.

glopesdev committed:
Use Doxyfile in diff mode for ease of reading

This will also make it easier to understand differences in Doxygen
defaults over time. XML and LaTeX targets were also removed.

glopesdev committed:
Use project README as main page

glopesdev committed:
Add continuous deployment criteria to docs README

glopesdev committed:
Exclude private members from public documentation
